Containerfile

```diff
@@ -10,13 +10,13 @@ WORKDIR /esi-leap
 
 # Copy the files of esi leap inside the container
 COPY run_services.sh run_services.sh
+COPY requirements.txt requirements.txt
 
 RUN mkdir -p /var/log/esi-leap
 RUN touch /var/log/esi-leap/esi-leap-dbsync.log
 
-# Run command to install dependencies for esi-leap
-RUN apt update -y && apt install -y python3-pymysql
-RUN pip3 install esi-leap mysql-connector pymysql
+# Install dependencies
+RUN pip install -r requirements.txt
 
 # Indicate that service listens on this port
 EXPOSE 7777
```

Comment:
You want to keep your `RUN pip install` operation as close as possible to the `COPY requirements.txt` operation: whenever you update `requirements.txt`, this invalidates the layer cache for every operation *after* the `COPY`. This means your container engine will need to re-run those steps as part of the build process rather than being able to utilize the cached layers, which increases the time required to build the image.

More broadly, to make the most effective use of the layer cache you want to order your operations from "least likely to change" to "most likely to change".

I might structure the file like this:

```
FROM docker.io/python:3.11

LABEL description="ESI-LEAP -- ESI lease policy manager"

# The working directory for esi-leap inside the container
WORKDIR /esi-leap

# Install dependencies
COPY requirements.txt requirements.txt
RUN pip install -r requirements.txt

RUN mkdir -p /var/log/esi-leap
RUN touch /var/log/esi-leap/esi-leap-dbsync.log

# Indicate that service listens on this port
EXPOSE 7777

# Copy the files of esi leap inside the container
COPY run_services.sh run_services.sh

CMD ["bash", "./run_services.sh"]
```

You can find more information about the Docker build cache [here](https://docs.docker.com/build/cache/). Podman (and buildah) operate on similar principles.

Reply:
Understood and will take care of it from next time.